clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,303 @@
+            return;
+        }
+        handleSuccessFindCoordinatorResponse(node, currentTimeMs);
+    }
+
+    public Node coordinator() {
+        return this.coordinator;
+    }
+
+    // Visible for testing
+    static class CoordinatorRequestState {
+        final static int RECONNECT_BACKOFF_EXP_BASE = 2;
+        final static double RECONNECT_BACKOFF_JITTER = 0.2;
+        private final ExponentialBackoff exponentialBackoff;
+        private long lastSentMs = -1;
+        private long lastReceivedMs = -1;
+        private int numAttempts = 0;
+        private long backoffMs = 0;
+
+        public CoordinatorRequestState(ConsumerConfig config) {
+            this.exponentialBackoff = new ExponentialBackoff(
+                    config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                    RECONNECT_BACKOFF_EXP_BASE,
+                    config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG),
+                    RECONNECT_BACKOFF_JITTER);
+        }
+
+        // Visible for testing
+        CoordinatorRequestState(final int reconnectBackoffMs,
+                                final int reconnectBackoffExpBase,
+                                final int reconnectBackoffMaxMs,
+                                final int jitter) {
+            this.exponentialBackoff = new ExponentialBackoff(
+                    reconnectBackoffMs,
+                    reconnectBackoffExpBase,
+                    reconnectBackoffMaxMs,
+                    jitter);
+        }
+
+        public void reset() {
+            this.lastSentMs = -1;
+            this.lastReceivedMs = -1;
+            this.numAttempts = 0;
+            this.backoffMs = exponentialBackoff.backoff(0);
+        }
+
+        public boolean canSendRequest(final long currentTimeMs) {
+            if (this.lastSentMs == -1) {
+                // no request has been sent
+                return true;
+            }
+
+            // TODO: I think there's a case when we want to resend the FindCoordinator when we haven't received
+            //  anything yet.
+            if (this.lastReceivedMs == -1 ||
+                    this.lastReceivedMs < this.lastSentMs) {
+                // there is an inflight request
+                return false;
+            }
+
+            return requestBackoffExpired(currentTimeMs);
+        }
+
+        public void updateLastSend(final long currentTimeMs) {
+            // Here we update the timer everytime we try to send a request. Also increment number of attempts.
+            this.lastSentMs = currentTimeMs;
+        }
+
+        public void updateLastFailedAttempt(final long currentTimeMs) {
+            this.lastReceivedMs = currentTimeMs;
+            this.backoffMs = exponentialBackoff.backoff(numAttempts);
+            this.numAttempts++;
+        }
+
+        private boolean requestBackoffExpired(final long currentTimeMs) {
+            return (currentTimeMs - this.lastReceivedMs) >= this.backoffMs;
```

Comment:
I think this overflows if `currentTimeMs` is less than `lastReceivedMs`. Could we just write this as this:
```java
return remainingBackoffMs(currentTimeMs) > 0;
```